openedx/core/djangoapps/programs/utils.py

```diff
@@ -127,6 +127,43 @@ def engaged_programs(self):
 
         return programs
 
+    @cached_property
+    def failed_course_runs(self):
+        """
+        Determine which course runs have been failed by the user.
+
+        Returns:
+            list of strings, each representing a certificate course run key
+        """
+        course_run_certificates = certificate_api.get_certificates_for_user(self.user.username)
+        return [
+            unicode(certificate['course_key']) for certificate in course_run_certificates
+            if not certificate_api.is_passing_status(certificate['status'])
```

Comment:
I see. Have you thought about ways to take advantage of the course_run_id/type dict returned by `reshape()`? You do require both of those values at different times. That structure or something like it may help you do what you're doing here more succinctly.